Utilize new APIs in the WebSocket implementation

PaulusParssinen opened the pull request:
Use few newer APIs in the managed websocket implementation. Gets rid of little manual endianness bit-twiddling that is common for code before `BinaryPrimitives`.

PaulusParssinen committed:
Use BinaryPrimitives in ManagedWebSocket

PaulusParssinen committed:
Use MemoryMarshal.CreateSpan instead of .ctor in WebSocketInflater

stephentoub commented on src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/WebSocketInflater.cs:
```diff
@@ -210,12 +211,12 @@ private void ReleaseBuffer()
-            byte b;
-            if (Inflate(stream, new Span<byte>(&b, 1), FlushCode.SyncFlush) == 0)
+            byte b = 0;
+            if (Inflate(stream, MemoryMarshal.CreateSpan(ref b, 1), FlushCode.SyncFlush) == 0)
```
This should just be `new Span(ref b) `

PaulusParssinen replied:
Right, Thanks! https://github.com/dotnet/runtime/pull/101953/commits/4708726602b0a364583010d4914440840531da23

PaulusParssinen committed:
Apply PR suggestion

PaulusParssinen commented on src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:
```diff
@@ -1495,9 +1490,7 @@ private static unsafe int ApplyMask(Span<byte> toMask, int mask, int maskIndex)
-                    int rolledMask = BitConverter.IsLittleEndian ?
-                        (int)BitOperations.RotateRight((uint)mask, maskIndex * 8) :
-                        (int)BitOperations.RotateLeft((uint)mask, maskIndex * 8);
+                    int rolledMask = (int)BitOperations.RotateRight((uint)mask, maskIndex * 8);
```
I was thinking of using `TensorPrimitives.Xor` here for the `sizeof(int)` elements but I think we can't guarantee alignment of the `toMask` for `MemoryMarshal.Cast` here..

PaulusParssinen committed:
Merge branch 'main' into use-new-apis-in-ws

PaulusParssinen committed:
Merge branch 'main' into use-new-apis-in-ws

PaulusParssinen commented:
Assuming the CI failures are unrelated but the fails were about missing log artifact directory, so I'm rolling the CI roulette one more time.

CarnaViire commented:
@zlatanov JFYI

CarnaViire commented on src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:
```diff
@@ -649,19 +650,13 @@ private static int WriteHeader(MessageOpcode opcode, byte[] sendBuffer, ReadOnly
-                    sendBuffer[i] = unchecked((byte)length);
-                    length /= 256;
-                }
+                BinaryPrimitives.WriteUInt64BigEndian(sendBuffer.AsSpan(2), (ulong)payload.Length);
```
We do already have a test that would verify it, is that correct?

zlatanov replied:
Not explicitly, but the deflate / inflate tests would hit this and verify its correctness.

PaulusParssinen replied:
The testing coverage is quite dangerously low but a lot of things would have breaked if these were wrong, for example the compression tests as said above😅

CarnaViire commented on src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:
```diff
@@ -649,19 +650,13 @@ private static int WriteHeader(MessageOpcode opcode, byte[] sendBuffer, ReadOnly
-                    sendBuffer[i] = unchecked((byte)length);
-                    length /= 256;
-                }
+                BinaryPrimitives.WriteUInt64BigEndian(sendBuffer.AsSpan(2), (ulong)payload.Length);
```
Also -- I believe there should be a place where it is read "back" -- shouldn't it be changed as well?

CarnaViire replied:
I believe it's here
https://github.com/dotnet/runtime/blob/9be628743487aac580a5e933f08d459dc50c80fb/src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs#L1157-L1173

zlatanov commented on src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:
```diff
@@ -1464,7 +1459,7 @@ private void ReleaseSendBuffer()
 
         private static int CombineMaskBytes(Span<byte> buffer, int maskOffset) =>
-            BitConverter.ToInt32(buffer.Slice(maskOffset));
+            BinaryPrimitives.ReadInt32LittleEndian(buffer.Slice(maskOffset));
```
I don't believe this is correct here. The mask is just 4 bytes which were packed as int in the old implementation just for storage.

zlatanov commented on src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:
```diff
@@ -1495,9 +1490,7 @@ private static unsafe int ApplyMask(Span<byte> toMask, int mask, int maskIndex)
-                    int rolledMask = BitConverter.IsLittleEndian ?
-                        (int)BitOperations.RotateRight((uint)mask, maskIndex * 8) :
-                        (int)BitOperations.RotateLeft((uint)mask, maskIndex * 8);
+                    int rolledMask = (int)BitOperations.RotateRight((uint)mask, maskIndex * 8);
```
I think we should keep the old implementation here, see my comment about the mask packing.

PaulusParssinen replied:
Sure, I was on the fence about this one.

PaulusParssinen committed:
Use BinaryPrimitives in more places

PaulusParssinen committed:
Restore the behavior where mask is stored in machine-endianness

* Implementation detail, both would work

PaulusParssinen committed:
Remove unused using directives

PaulusParssinen committed:
Read the payload length as unsigned for 16-bit case

RFC6455, Section 5.2. quote: "If 126, the following 2 bytes interpreted as a 16-bit unsigned integer are the payload length."

PaulusParssinen committed:
Consolidate some validation logic to WebSocketValidata.cs

* Use CallerArgumentExpression to flow the parameterName in ValidateArraySegment

PaulusParssinen committed:
Remove unused ValidateBuffer

PaulusParssinen committed:
Use new throw helpers in ValueWebSocketReceiveResult

PaulusParssinen committed:
Use ValueTask.FromException

PaulusParssinen commented:
I just realized that `WebSocketValidate.cs` is in shared folder, but.. it's not actually imported by any of the projects in the class and only other place it's "synced" to is azure-relay-dotnet?

https://grep.app/search?q=websocketvalidate&filter[lang][0]=C%23&filter[path][0]=src/&filter[repo][0]=Azure/azure-relay-dotnet

Can we move the class to `System.Net.WebSockets`?

PaulusParssinen committed:
Revert "Remove unused ValidateBuffer"

This reverts commit 053f0ac0f97ba52216b61652f4fe72d2e074b68c.

PaulusParssinen committed:
Oops

PaulusParssinen committed:
Fix ArgumentNullException for ValidateArraySegment

stephentoub commented on src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs:
```diff
@@ -116,22 +138,20 @@ internal static void ValidateCloseStatus(WebSocketCloseStatus closeStatus, strin
-            {
-                throw new ArgumentOutOfRangeException(parameterName + "." + nameof(arraySegment.Count));
-            }
+            string arrayParameterName = parameterName + "." + nameof(arraySegment.Array);
```
This is not an acceptable change. It is now allocating a string regardless of whether the argument is nulll.

PaulusParssinen replied:
Ah right! Thanks for spotting. Will revert this.

PaulusParssinen replied:
Custom interpolated string handler short-circuit would be nice in the new throw helpers for this 😅

stephentoub replied:
> Not worth it probably.

No. There would still be additional checks performed. In the production path where the argument should _never_ be null, we want the overhead to be nothing more than a single comparison/branch.

stephentoub commented on src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs:
```diff
@@ -37,7 +47,19 @@ internal static partial class WebSocketValidate
+            => ThrowIfInvalidState(currentState, isDisposed, s_validCloseOutputStates);
+
+        internal static void ThrowIfInvalidCloseState(WebSocketState currentState, bool isDisposed)
+            => ThrowIfInvalidState(currentState, isDisposed, s_validCloseStates);
```
Why is adding all these methods beneficial?

PaulusParssinen replied:
``WebSocketValidate.ThrowIfInvalidState(currentState, isDisposed, WebSocketValidate.ValidCloseStates);`` felt too long winded. I don't feel strongly about these, will revert 👍

PaulusParssinen committed:
Revert "Consolidate some validation logic to WebSocketValidata.cs"

This reverts commit c515ba99c2bb25a7dd376eb66ab50f4425ab1203.

PaulusParssinen committed:
Revert ValueWebSocketReceiveResult.ThrowMessageTypeOutOfRange

zlatanov commented on src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketReceiveResult.cs:
```diff
@@ -1,8 +1,6 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System.Diagnostics;
```
I would abstain from removing unused namespaces from files in this PR as it is not related, and especially when it's the only change in the file.

zlatanov commented on src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketException.cs:
```diff
@@ -2,7 +2,6 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.ComponentModel;
-using System.Diagnostics.CodeAnalysis;
```
See my other comment about namespaces.

zlatanov commented on src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketCloseStatus.cs:
```diff
@@ -1,8 +1,6 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System.Diagnostics.CodeAnalysis;
```
See my other comment about namespaces.

PaulusParssinen committed:
Revert "Remove unused using directives"

This reverts commit 820dc0491c383bd0d1d87b3a08433f417045ddb1.

zlatanov commented:
LGTM. Thanks!

CarnaViire reviewed (approved):
LGTM, thanks! (and sorry for the delay)
